### PR TITLE
feature: tracking app metrics for web olny

### PR DESCRIPTION
**Pull request by mydiemho**
1. tracking app version with every metric
1. tracking router/page changes
1. tracking exception
    * include error type and message
    * uncaught exception are handled automatically by AppInsights
    * caught exception are handle manually
1. tracking redux actions
    * only tracking action type

To see what the data looks like, checkout [this query](https://ms.portal.azure.com#@72f988bf-86f1-41af-91ab-2d7cd011db47/blade/Microsoft_OperationsManagementSuite_Workspace/AnalyticsBlade/initiator/AnalyticsShareLinkToQuery/isQueryEditorVisible/true/scope/%7B%22resources%22%3A%5B%7B%22resourceId%22%3A%22%2Fsubscriptions%2Fd36d0808-a967-4f73-9fdc-32ea232fc81d%2FresourceGroups%2Fmyho-development%2Fproviders%2Fmicrosoft.insights%2Fcomponents%2Fmyho-appInsights%22%7D%5D%7D/query/H4sIAAAAAAAAA0suLS7Jz3UtS80rKeaqUcjJzM0sUTA14AIAetTGjBgAAAA%3D/isQueryBase64Compressed/true/timespanInIsoFormat/P1D)

[AB#17694](https://dev.azure.com/dwrdev/web/wi.aspx?pcguid=80f9922c-b416-454c-92c6-ab7b6867e0b2&id=17694)

### CODE REVIEW

File: src/redux/middleware/appInsights.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { AnyAction, Dispatch, Middleware, MiddlewareAPI } from "redux";
+import { trackReduxAction } from "../../telemetry";
+
+export function createAppInsightsLogger(): Middleware {
+    return (store: MiddlewareAPI<Dispatch<AnyAction>>) => (next: Dispatch<AnyAction>) => (action: any) => {
+        trackReduxAction(action);
```
Comment (pjlittle) [inline src/redux/middleware/appInsights.ts]: this is great :boom:

File: src/telemetry.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import history from "./history";
+import { ApplicationInsights, SeverityLevel } from "@microsoft/applicationinsights-web";
+import { version } from "../package.json";
+import { isElectron } from "./common/hostProcess";
+import { Action } from "redux";
+import { ErrorCode } from "./models/applicationState";
+import { Error } from "tslint/lib/error";
+
+// vott-app-insights
+let INSTRUMENTATION_KEY = "0b9e5117-c78d-40c9-9338-921092cde49a";
```
Comment (pjlittle) [inline src/telemetry.ts]: thinking we should dynamically fill this in via Web pack.  It's just an identifier though.

**Comment by pjlittle** [comment]
LGTM - I'd be interested to see what the reporting/dashboard side looks like.

File: package.json [hunk 3]
```diff
@@ -27,6 +27,7 @@
         "node-int64": "^0.4.0",
         "rc-menu": "^7.4.21",
         "react": "^16.7.0",
+        "react-appinsights": "^3.0.0-rc.5",
```
Comment (pjlittle) [inline package.json]: looks like they're close to releasing v3? Did you run into any weirdness w/ this RC?

Comment (mydiemho) [reply package.json]: no, haven't notice anything strange

File: src/telemetry.ts (hunk 2 above)
Comment (mydiemho) [reply src/telemetry.ts]: moving the key to `.env*`

File: src/App.test.tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ import createReduxStore from "./redux/store/store";
 import initialState from "./redux/store/initialState";
 import { IApplicationState } from "./models//applicationState";
 import { mount } from "enzyme";
-import { HashRouter } from "react-router-dom";
+import { HashRouter, Router } from "react-router-dom";
```
Comment (wbreza) [inline src/App.test.tsx]: HashRouter not used anymore and we should be able to remove?

File: src/redux/middleware/appInsights.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { AnyAction, Dispatch, Middleware, MiddlewareAPI } from "redux";
+import { trackReduxAction } from "../../telemetry";
+
+export function createAppInsightsLogger(): Middleware {
```
Comment (wbreza) [inline src/redux/middleware/appInsights.ts]: Add JS docs to describe this middleware

File: src/telemetry.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+if (Env.get() !== "production") {
+    // for development/testing
+    // myho-appinsights
+    debug = true;
+    maxBatchSize = 0;  // send telemetry as soon as it's collected
+}
+
+let appInsights;
+
+export function setUpAppInsights() {
```
Comment (wbreza) [inline src/telemetry.ts]: Add JS docs to these exported functions

File: src/App.test.tsx (hunk 4 above)
Comment (mydiemho) [reply src/App.test.tsx]: yeah, it got added back with rebasing, will remove

**Pull request merged**